sycl/doc/design/SYCL-1Pass.md

```diff
@@ -0,0 +1,57 @@
+# SYCL 1-Pass Compilation Model
+
+Currently the DPC++ compiler utilizes a scheme which needs multiple invocations of the compiler.  Specifically, it invokes the cc1 compiler once for each device compilation and one more time for the host compilation.  This document describes design changes necessary (in the various components of the DPC++ Toolchain) to switch to a single invocation of the cc1 compiler for each Translation Unit (TU).
+
+## Compiler Front End (FE) Changes
+
+This design reuqires some changes from the FE.  In particular, it needs to emit code via the normal/standard C++ rules, and in addition, needs to emit functions that are required for Kernel code, which may not have been emitted via the normal C++ rules.
+
+### Generation and dispatch to the OpenCL Kernel
+
+The FE will generate the OpenCL Kernel which creates the Functor object that it will call into.  FE then replaces the [[sycl_kernel]] function with one that dispatches to the OpenCL Kernel.  New APIs from the SYCL Runtime Library (RTL) are needed to accomplish this task.
```

Comment:
The key here is that the sycl_kernel function is actually code-gen'ed 2x.  Again, all of this work happens in clang-codegen now, instead of in SemaSYCL.  Essentially, replace the normal "EmitFunctionDecl" code for the sycl_kernel functions with 2 separate functions generated: 

The host-side, which generates the decomposition of the function object, does the calls into the RTL to move this data over to the device, and finally calls the OpenCL kernel.

The Device-side, which is just the open-cl kernel, similar to how we already generate for device side.  HOWEVER, this now needs to happen at the IR level instead of the AST/Sema level.